_posts/2022-06-18-zio-streams.md

```diff
@@ -0,0 +1,537 @@
+    ZStream("0", "1", "2", "3", "4", "5", "6", "7", "8", "9")
+
+  // We can use contramap to use our take5 logic, and operate on stringStream with it.
+  val take5Strings: ZSink[Any, Nothing, String, Int, Chunk[String]] =
+    take5Map.contramap[String](_.toInt)
+
+// This is equivalent to take5Strings above
+  val take5Dimap: ZSink[Any, Nothing, String, Int, Chunk[String]] =
+    take5.dimap[String, Chunk[String]](_.toInt, _.map(_.toString))
+```
+
+Notice that he `L` is still `Int`. With contramap, we are operating on input
+element to get it to the correct type (`String`), however, if anything is left
+over, then it wasn't operated on - meaning it's still the input type from
+`take5` (`Int`).
+
+### ZPipelines
+
+A ZPipeline converts one ZStream to another ZStream. It can conceptually be
+thought of as
+`type ZPipeline[Env, Err, In, Out] = ZStream[Env, Err, In] => ZStream[Env, Err, Out]`.
+
+The main use case of a ZPipeline is to separate out reusable transform logic,
+that can then be composed together with other ZPipelines, and then placed in
+between any appropriate ZStream and a ZSink.
+
+From our example above, let's say we wanted to sum the stream of Strings.
+Instead of adapting our `sum` ZSink with contramap, we can write:
+
+```scala
+  //stringStream.run(sum) <- This doesn't compile
+
+  val businessLogic: ZPipeline[Any, Nothing, String, Int] =
+    ZPipeline.map[String, Int](_.toInt)
+
+  val zio: ZIO[Any, Nothing, Int] =
+    stringStream.via(businessLogic).run(sum)
+```
+
+Along with the typical collection like operations you'd expect, there are a
+number of addition ones that are available directly on `ZStream` and can be
+referenced in the
+[official docs](https://zio.dev/next/datatypes/stream/zstream/#operations) .
+
+## Handling Failures
+
+If we think of some `val s1: ZStream[R, E, A]`, we can use the `orElse` methods
+to directly provide a recovery stream. The interesting thing of `orElseEither`
+is that we can distinguish the transition when `s1` failed, and `s2` was used,
+as `s1` values would be a `Left` and values from `s2` would be `Right`.
+
+```scala
+s1.orElse(s2: => ZStream[R1, E1, A1]): ZStream[R1, E1, A1]
+s1.orElseEither(s2: => ZStream[R1, E1, A1]): ZStream[R1, E1, Either[A, A1]]
+```
+
+If we want to recover from specific failures via
+`PartialFunction[E, ZStream[R1, E1, A1]]`s we can use `catchSome` ( or
+`catchSomeCause` for causes - a.k.a _defects_).
+
+```scala
+s1.catchSome {
+  case e1: Error1 => s2
+}
+```
+
+If we want to exhaustively recover from errors via `E => ZStream[R1, E2, A1]`)
+(and causes), we can use `catchAll` (or `catchAllCause`).
+
+```scala
+s1.catchAll {
+  case e1: Error1 => s2
+  case e2: Error2 => s3
+  case _          => s4
+}
+```
+
+If we want to push our error handling downwards, we can also transform a
+`ZStream[R, E, A]` to `ZStream[R, Nothing, Either[E, A]]` via `either` (e.g.
+`s1.either`).
+
+Recovering form a failure in a `ZStream` generally revolves around providing a
```

Comment:
I didn't understand too much of this paragraph. An example of in-place dropping and collectRight will clear all doubts.

Reply:
Yes, this wasn't very clear at all. I've reworked it along with the better examples of error handling above it.